pkg/endpoint/policy.go

```
@@ -267,12 +268,66 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
 	// DistillPolicy converts a SelectorPolicy in to an EndpointPolicy
 	stats.endpointPolicyCalculation.Start()
 	result.endpointPolicy = selectorPolicy.DistillPolicy(e.getLogger(), e, desiredRedirects)
+	if pcr.NeedsRelease {
+		oldPolicy.Detach()
+	}
 	stats.endpointPolicyCalculation.End(true)
 
 	datapathRegenCtxt.policyResult = result
 	return nil
 }
 
+func (e *Endpoint) waitForPolicyComputationResult(
```

It would be useful to have a guarantee that whenever a new or current selector policy is returned (via `Result`), the caller could rely on that policy not being detached while this endpoint is preparing to use it.

This ties in with the other comment on not detaching on the policy change, but on the last user.

Even the "last user" detachment policy has a potential race: What it the last *current* user stops using the policy, while a new endpoint has just received the policy (via this function) and is preparing to start using it? Currently we let the policy attach fail silently, leading to the new endpoint using a detached selector policy and not getting any incremental updates until the endpoint is regenerated again. The minimum we should do to fix this is to return an error up from the `EndpointPolicy.attach` that tries to attach to a detached selector policy, and have the endpoint retry the policy computation immediately.